File: graph-worker/operational-workers/internal/common/utils/metrics.go
package utils

import (
	"fmt"

	"github.com/prometheus/client_golang/prometheus"
)

// ProcessorMetrics provides a standard set of metrics for message processors
type ProcessorMetrics struct {
	ProcessingTime    prometheus.Histogram
	ProcessingErrors  prometheus.Counter
	ProcessingSuccess prometheus.Counter
	MessagesInFlight  prometheus.Gauge
}

// NewProcessorMetrics creates a new set of processor metrics
func NewProcessorMetrics(workerType string) *ProcessorMetrics {
	metrics := &ProcessorMetrics{
		ProcessingTime: prometheus.NewHistogram(
			prometheus.HistogramOpts{
				Name:    fmt.Sprintf("%s_processing_time_seconds", workerType),
				Help:    fmt.Sprintf("Time taken to process messages for %s worker", workerType),
				Buckets: prometheus.DefBuckets, // 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10
			},
		),
		ProcessingErrors: prometheus.NewCounter(
			prometheus.CounterOpts{
				Name: fmt.Sprintf("%s_processing_errors_total", workerType),
				Help: fmt.Sprintf("Total number of processing errors for %s worker", workerType),
			},
		),
		ProcessingSuccess: prometheus.NewCounter(
			prometheus.CounterOpts{
				Name: fmt.Sprintf("%s_processing_success_total", workerType),
				Help: fmt.Sprintf("Total number of successful processing for %s worker", workerType),
			},
		),
		MessagesInFlight: prometheus.NewGauge(
			prometheus.GaugeOpts{
				Name: fmt.Sprintf("%s_messages_in_flight", workerType),
				Help: fmt.Sprintf("Number of messages currently being processed by %s worker", workerType),
			},
		),
	}

	// Register metrics
	prometheus.MustRegister(
		metrics.ProcessingTime,
		metrics.ProcessingErrors,
		metrics.ProcessingSuccess,
		metrics.MessagesInFlight,
	)

	return metrics
}

// RecordProcessingStart records the start of message processing
func (m *ProcessorMetrics) RecordProcessingStart() {
	m.MessagesInFlight.Inc()
}

// RecordProcessingSuccess records successful message processing
func (m *ProcessorMetrics) RecordProcessingSuccess() {
	m.ProcessingSuccess.Inc()
	m.MessagesInFlight.Dec()
}

// RecordProcessingError records failed message processing
func (m *ProcessorMetrics) RecordProcessingError() {
	m.ProcessingErrors.Inc()
	m.MessagesInFlight.Dec()
}

// StartTimer returns a timer for recording processing duration
func (m *ProcessorMetrics) StartTimer() *prometheus.Timer {
	return prometheus.NewTimer(m.ProcessingTime)
}

File: graph-worker/graphrag-service/src/monitoring/metrics.py
import time
from contextlib import contextmanager

from prometheus_client import Counter, Gauge, Histogram


class PrometheusMetrics:
    """Prometheus metrics for GraphRAG worker."""

    def __init__(self) -> None:
        self.messages_processed = Counter(
            "graphrag_messages_processed_total",
            "Total messages processed",
            ["status"],
        )

        self.documents_processed = Counter(
            "graphrag_documents_processed_total",
            "Total documents processed",
        )

        self.processing_duration = Histogram(
            "graphrag_processing_duration_seconds",
            "Document processing duration",
            buckets=[60, 120, 300, 600, 900, 1200, 1800, 3600],
        )

        self.current_processing = Gauge(
            "graphrag_currently_processing",
            "Number of documents currently being processed",
        )

        self.last_success_timestamp = Gauge(
            "graphrag_last_success_timestamp",
            "Timestamp of last successful processing",
        )

    def record_success(self) -> None:
        self.messages_processed.labels(status="success").inc()
        self.documents_processed.inc()
        self.last_success_timestamp.set_to_current_time()

    def record_error(self, error_type: str) -> None:
        self.messages_processed.labels(status=f"error_{error_type}").inc()

    @contextmanager
    def track_duration(self):
        self.current_processing.inc()
        start = time.time()
        try:
            yield
        finally:
            duration = time.time() - start
            self.processing_duration.observe(duration)
            self.current_processing.dec()
